packages/wrangler/src/deployment-bundle/bundle.ts

```diff
@@ -371,18 +385,29 @@ export async function bundleWorker(
 		...(jsxFactory && { jsxFactory }),
 		...(jsxFragment && { jsxFragment }),
 		...(tsconfig && { tsconfig }),
-		watch,
 		// The default logLevel is "warning". So that we can rewrite errors before
 		// logging, we disable esbuild's default logging, and log build failures
 		// ourselves.
 		logLevel: "silent",
 	};
-	let result;
+
+	let result: esbuild.BuildResult<typeof buildOptions>;
+	let stop: BundleResult["stop"];
 	try {
-		result = await esbuild.build(buildOptions);
+		if (watch) {
+			const ctx = await esbuild.context(buildOptions);
+			await ctx.watch();
+			result = await initialBuildResultPromise;
```

Comment:
Yes since `onEnd` will still be run and then [this](https://github.com/cloudflare/workers-sdk/pull/3772/files#diff-08737a62058e3e6d5614a637f6cbdf460b6238a2f749bc2d30de3756f8463e51R116) will be run

Reply:
Tested manually to double check